In this kernel we will study how to do a topic modelling on a Given Dataset.Please Vote if you like the work

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/topic-modeling/npr.csv


In [2]:
import pandas as pd

In [3]:
npr=pd.read_csv('../input/topic-modeling/npr.csv')
npr.head()

Article
0  In the Washington of 2016, even when the polic...
1    Donald Trump has used Twitter  —   his prefe...
2    Donald Trump is unabashedly praising Russian...
3  Updated at 2:50 p. m. ET, Russian President Vl...
4  From photography, illustration and video, to d...

In [4]:
npr.shape

(11992, 1)

We can see that our dataset has 11992 documents.Our task will be to segegrate this documents into various topics

**Displaying the first document **

In [5]:
npr['Article'][0]

'In the Washington of 2016, even when the policy can be bipartisan, the politics cannot. And in that sense, this year shows little sign of ending on Dec. 31. When President Obama moved to sanction Russia over its alleged interference in the U. S. election just concluded, some Republicans who had long called for similar or more severe measures could scarcely bring themselves to approve. House Speaker Paul Ryan called the Obama measures ”appropriate” but also ”overdue” and ”a prime example of this administration’s ineffective foreign policy that has left America weaker in the eyes of the world.” Other GOP leaders sounded much the same theme. ”[We have] been urging President Obama for years to take strong action to deter Russia’s worldwide aggression, including its   operations,” wrote Rep. Devin Nunes,  . chairman of the House Intelligence Committee. ”Now with just a few weeks left in office, the president has suddenly decided that some stronger measures are indeed warranted.” Appearing 

In [6]:
len(npr)

11992

**Preprocessing of the data**

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

In [8]:
cv=CountVectorizer(max_df=0.9,min_df=2,stop_words='english')

In [9]:
dtm=cv.fit_transform(npr['Article'])

In [10]:
dtm

<11992x54777 sparse matrix of type '<class 'numpy.int64'>'
	with 3033388 stored elements in Compressed Sparse Row format>

In [11]:
from sklearn.decomposition import LatentDirichletAllocation

In [12]:
LDA=LatentDirichletAllocation(n_components=7,random_state=42)

In [13]:
LDA.fit(dtm)

LatentDirichletAllocation(batch_size=128, doc_topic_prior=None,
                          evaluate_every=-1, learning_decay=0.7,
                          learning_method='batch', learning_offset=10.0,
                          max_doc_update_iter=100, max_iter=10,
                          mean_change_tol=0.001, n_components=7, n_jobs=None,
                          perp_tol=0.1, random_state=42, topic_word_prior=None,
                          total_samples=1000000.0, verbose=0)

**Grab the vocubalary of words**

In [14]:
len(cv.get_feature_names())

54777

In [15]:
type(cv.get_feature_names())

list

In [16]:
cv.get_feature_names()[41000]

'reproductive'

In [17]:
import random
random_word_id=random.randint(0,54777)
cv.get_feature_names()[random_word_id]

'rubicon'

**Grab the topics**

In [18]:
len(LDA.components_)

7

In [19]:
LDA.components_.shape

(7, 54777)

In [20]:
LDA.components_

array([[8.64332806e+00, 2.38014333e+03, 1.42900522e-01, ...,
        1.43006821e-01, 1.42902042e-01, 1.42861626e-01],
       [2.76191749e+01, 5.36394437e+02, 1.42857148e-01, ...,
        1.42861973e-01, 1.42857147e-01, 1.42906875e-01],
       [7.22783888e+00, 8.24033986e+02, 1.42857148e-01, ...,
        6.14236247e+00, 2.14061364e+00, 1.42923753e-01],
       ...,
       [3.11488651e+00, 3.50409655e+02, 1.42857147e-01, ...,
        1.42859912e-01, 1.42857146e-01, 1.42866614e-01],
       [4.61486388e+01, 5.14408600e+01, 3.14281373e+00, ...,
        1.43107628e-01, 1.43902481e-01, 2.14271779e+00],
       [4.93991422e-01, 4.18841042e+02, 1.42857151e-01, ...,
        1.42857146e-01, 1.43760101e-01, 1.42866201e-01]])

**Grab Highest Probability word for each topic**

In [21]:
single_topic=LDA.components_[0]

In [22]:
single_topic.argsort()

array([ 2475, 18302, 35285, ..., 22673, 42561, 42993])

In [23]:
import numpy as np
arr=np.array([10,200,1])

In [24]:
arr

array([ 10, 200,   1])

In [25]:
arr.argsort()

array([2, 0, 1])

In [26]:
# Argsort gives index in ascending order 
single_topic.argsort()[-10:]

array([33390, 36310, 21228, 10425, 31464,  8149, 36283, 22673, 42561,
       42993])

In [27]:
top_ten_words=single_topic.argsort()[-10:]

In [28]:
for index in top_ten_words:
    print(cv.get_feature_names()[index])

new
percent
government
company
million
care
people
health
said
says


In [29]:
for i,topic in enumerate(LDA.components_):
    print(f"THE TOP 15 WORDS FOR THE TOPIC #{i}")
    print([cv.get_feature_names()[index] for index in topic.argsort()[-15:]])
    print('\n')
    print('\n')

THE TOP 15 WORDS FOR THE TOPIC #0
['companies', 'money', 'year', 'federal', '000', 'new', 'percent', 'government', 'company', 'million', 'care', 'people', 'health', 'said', 'says']




THE TOP 15 WORDS FOR THE TOPIC #1
['military', 'house', 'security', 'russia', 'government', 'npr', 'reports', 'says', 'news', 'people', 'told', 'police', 'president', 'trump', 'said']




THE TOP 15 WORDS FOR THE TOPIC #2
['way', 'world', 'family', 'home', 'day', 'time', 'water', 'city', 'new', 'years', 'food', 'just', 'people', 'like', 'says']




THE TOP 15 WORDS FOR THE TOPIC #3
['time', 'new', 'don', 'years', 'medical', 'disease', 'patients', 'just', 'children', 'study', 'like', 'women', 'health', 'people', 'says']




THE TOP 15 WORDS FOR THE TOPIC #4
['voters', 'vote', 'election', 'party', 'new', 'obama', 'court', 'republican', 'campaign', 'people', 'state', 'president', 'clinton', 'said', 'trump']




THE TOP 15 WORDS FOR THE TOPIC #5
['years', 'going', 've', 'life', 'don', 'new', 'way', 'music', 

In [30]:
topic_results=LDA.transform(dtm)

In [31]:
topic_results.shape

(11992, 7)

In [32]:
topic_results[0]

array([1.61040465e-02, 6.83341493e-01, 2.25376318e-04, 2.25369288e-04,
       2.99652737e-01, 2.25479379e-04, 2.25497980e-04])

In [33]:
topic_results[0].round(2)

array([0.02, 0.68, 0.  , 0.  , 0.3 , 0.  , 0.  ])

In [34]:
npr['Article'][0]

'In the Washington of 2016, even when the policy can be bipartisan, the politics cannot. And in that sense, this year shows little sign of ending on Dec. 31. When President Obama moved to sanction Russia over its alleged interference in the U. S. election just concluded, some Republicans who had long called for similar or more severe measures could scarcely bring themselves to approve. House Speaker Paul Ryan called the Obama measures ”appropriate” but also ”overdue” and ”a prime example of this administration’s ineffective foreign policy that has left America weaker in the eyes of the world.” Other GOP leaders sounded much the same theme. ”[We have] been urging President Obama for years to take strong action to deter Russia’s worldwide aggression, including its   operations,” wrote Rep. Devin Nunes,  . chairman of the House Intelligence Committee. ”Now with just a few weeks left in office, the president has suddenly decided that some stronger measures are indeed warranted.” Appearing 

In [35]:
topic_results[0].argmax()

1

In [36]:
npr['Topic']=topic_results.argmax(axis=1)

In [37]:
npr

Article  Topic
0      In the Washington of 2016, even when the polic...      1
1        Donald Trump has used Twitter  —   his prefe...      1
2        Donald Trump is unabashedly praising Russian...      1
3      Updated at 2:50 p. m. ET, Russian President Vl...      1
4      From photography, illustration and video, to d...      2
...                                                  ...    ...
11987  The number of law enforcement officers shot an...      1
11988    Trump is busy these days with victory tours,...      4
11989  It’s always interesting for the Goats and Soda...      3
11990  The election of Donald Trump was a surprise to...      4
11991  Voters in the English city of Sunderland did s...      0

[11992 rows x 2 columns]

In [38]:
mytopic_dict={0:'Health',1:'Election',2:'Family',3:'Politics',4:'Election',5:'Music',6:'Education'}
npr['Topic Label']=npr['Topic'].map(mytopic_dict)

In [39]:
npr

Article  Topic Topic Label
0      In the Washington of 2016, even when the polic...      1    Election
1        Donald Trump has used Twitter  —   his prefe...      1    Election
2        Donald Trump is unabashedly praising Russian...      1    Election
3      Updated at 2:50 p. m. ET, Russian President Vl...      1    Election
4      From photography, illustration and video, to d...      2      Family
...                                                  ...    ...         ...
11987  The number of law enforcement officers shot an...      1    Election
11988    Trump is busy these days with victory tours,...      4    Election
11989  It’s always interesting for the Goats and Soda...      3    Politics
11990  The election of Donald Trump was a surprise to...      4    Election
11991  Voters in the English city of Sunderland did s...      0      Health

[11992 rows x 3 columns]